fms_to_hf.py

```diff
@@ -98,7 +98,9 @@ def convert_to_hf(model: LLaMA) -> LlamaForCausalLM:
 def main(model_variant, compiled, load_path, save_path, tokenizer_name_or_path):
     print("Initializing model...")
     llama_config = get_model_config(model_variant)
-    model = LLaMA(llama_config, orig_init=True)
+    with torch.device("meta"):
```

Comment:
Also, what's actually being saved? I believe we've removed the call to `reset_parameters` from the LLaMA constructor

Reply:
@nairbv this part is inside the fms->hf conversion script.   It basically create an empty sketch of the model, and load saved checkpoints from the checkpoints, and write out the HF format.    Here, specifically, we just need an way (any way) to create the sketch (holder) for the to-be-loaded checkpoint.    There is no `reset_parameters` involved as we are not doing any model init here.

Reply:
right, I'm just asking why instantiating the model with device=meta and then calling to_empty is faster than simply instantiating the model on cpu?

Reply:
the description of the PR is that it's because "we don't care about the parameter init" ... if "parameter init" means "reset_parameters" I believe that's not getting called anymore so shouldn't change the speed?

 I guess you mean something else but I'm not sure what... we might also want a comment on this saying what it's for

Reply:
@nairbv ah. I misunderstood the question.

so, basically:
`to_emtpy` is a materialized-but-uninitialized form, and it is much faster than a true init, as true init does a "true init" (e.g. by default, each module has a default/build-in init schema, like drawing numbers from a uniform distribution).

on high level we can think there are three types:
non-materialized (meta device tensors): fastest.  but since it is not materized (no storage attached, nothing), you cannot load checkpoint to such tensors/modules.
materialized-but-uninitialized:  typically useful as a placeholder when you know for sure you will "init it later", e.g. either load a checkpoint, or apply a custom init later.
materialized-and-initialized: most cases.

Reply:
`we don't care about the parameter init` -> `we don't care about initial paramemter init, as it will be overwritten by checkpoints`

Reply:
ah I see. though we're not calling reset_parameters, each individual module is (https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/linear.py#L105).

it might be worth adding a comment since the value of this meta device may not be obvious

Reply:
@nairbv yes, exactly. there is always default init there, and what we are doing here is to avoid ANY sort of init - we only create an empty place holder using `empty`, which only gives the tensor/module a storage (meta device is shape only, while empty is shape+storage-only).  And the values are not initialized thus it is a no-op.